resource/resource_graph_export.go

```diff
@@ -11,6 +12,61 @@ import (
 	"go.viam.com/rdk/utils"
 )
 
+const snapshotLimit = 500
+
+// Visualizer stores a history resource graph DOT snapshots.
+type Visualizer struct {
+	snapshots list.List
+}
+
+// SnapshotInfo contains a DOT snapshot string along with metadata.
+type SnapshotInfo struct {
+	Dot   string
+	Index int
+	Count int
+}
+
+// SaveSnapshot takes a DOT snapshot of a resource graph.
+func (viz *Visualizer) SaveSnapshot(g *Graph) error {
+	snapshot, err := g.ExportDot()
+	if err != nil {
+		return err
+	}
+	latestSnapshot := viz.snapshots.Front()
+	if latestSnapshot != nil && snapshot == latestSnapshot.Value.(string) {
+		// Nothing changed since the last snapshot
+		return nil
+	}
+	viz.snapshots.PushFront(snapshot)
+	if viz.snapshots.Len() > snapshotLimit {
+		viz.snapshots.Remove(viz.snapshots.Back())
+	}
+	return nil
+}
+
+// Count returns the number of snapshots currents stored.
+func (viz *Visualizer) Count() int { return viz.snapshots.Len() }
+
+// GetSnapshot returns a DOT snapshot at a given index, where index 0 is the latest
+// snapshot.
+func (viz *Visualizer) GetSnapshot(index int) (SnapshotInfo, error) {
+	result := SnapshotInfo{Index: index, Count: viz.snapshots.Len()}
+
+	if result.Count == 0 {
+		return result, errors.New("no snapshots")
+	}
+	if index < 0 || index >= result.Count {
+		return result, errors.New("out of range")
+	}
+	snapshot := viz.snapshots.Front()
+	for index > 0 && snapshot.Next() != nil {
```

Comment:
> I can't tell if the Next() != nil check is "belt and suspenders" (i.e: we checked the count is in range so Next() will get a real element -- but "just in case it doesn't"). Or if the intention was that without locks, a new snapshot can be added while iterating (invalidating the count check).

"just in case" was the intention, but you're right that this can lead to inconsistent results. returning the actual index with an explanation is a nice simple improvement, I'll do that for now

Reply:
on second thought, maybe it's clearer to just error and handle appropriately downstream